File: modules/zig/src/main/java/com/falsepattern/zigbrains/zig/editing/ZigEnterInQuotedStringHandler.java
package com.falsepattern.zigbrains.zig.editing;

import com.falsepattern.zigbrains.zig.parser.ZigFile;
import com.falsepattern.zigbrains.zig.psi.ZigStringLiteral;
import com.falsepattern.zigbrains.zig.psi.ZigTypes;
import com.falsepattern.zigbrains.zig.util.PsiTextUtil;
import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegateAdapter;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.actionSystem.EditorActionHandler;
import com.intellij.openapi.util.Ref;
import com.intellij.psi.PsiFile;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.jetbrains.annotations.NotNull;

@RequiredArgsConstructor
public class ZigEnterInQuotedStringHandler extends EnterHandlerDelegateAdapter {
    @Override
    public Result preprocessEnter(@NotNull PsiFile file,
                                  @NotNull Editor editor,
                                  @NotNull Ref<Integer> caretOffsetRef,
                                  @NotNull Ref<Integer> caretAdvanceRef,
                                  @NotNull DataContext dataContext,
                                  EditorActionHandler originalHandler) {
        if (!(file instanceof ZigFile)) {
            return Result.Continue;
        }

        val caretOffset = (int)caretOffsetRef.get();
        var psiAtOffset = file.findElementAt(caretOffset);
        if (psiAtOffset instanceof LeafPsiElement leaf) {
            if ( leaf.getTokenType() == ZigTypes.STRING_LITERAL_SINGLE) {
                psiAtOffset = leaf.getParent();
            }
        }
        if (psiAtOffset instanceof ZigStringLiteral str &&
            !str.isMultiLine() &&
            str.getTextOffset() < caretOffset) {
            PsiTextUtil.splitString(editor, str, caretOffset, true);
            return Result.Stop;
        }
        return Result.Continue;
    }
}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/zig/editing/ZigEnterInQuotedStringHandler.kt
package com.falsepattern.zigbrains.zig.editing

import com.falsepattern.zigbrains.zig.psi.ZigFile
import com.falsepattern.zigbrains.zig.psi.ZigStringLiteral
import com.falsepattern.zigbrains.zig.psi.ZigTypes
import com.falsepattern.zigbrains.zig.psi.splitString
import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegate.Result
import com.intellij.codeInsight.editorActions.enter.EnterHandlerDelegateAdapter
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.actionSystem.EditorActionHandler
import com.intellij.openapi.util.Ref
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.source.tree.LeafPsiElement

class ZigEnterInQuotedStringHandler: EnterHandlerDelegateAdapter() {
    override fun preprocessEnter(
        file: PsiFile,
        editor: Editor,
        caretOffset: Ref<Int>,
        caretAdvance: Ref<Int>,
        dataContext: DataContext,
        originalHandler: EditorActionHandler?
    ): Result {
        if (file !is ZigFile)
            return Result.Continue

        val caretOffset = caretOffset.get()!!
        var psiAtOffset = file.findElementAt(caretOffset) ?: return Result.Continue
        if (psiAtOffset is LeafPsiElement) {
            if (psiAtOffset.elementType == ZigTypes.STRING_LITERAL_SINGLE) {
                psiAtOffset = psiAtOffset.parent ?: return Result.Continue
            }
        }
        if (psiAtOffset is ZigStringLiteral &&
            !psiAtOffset.isMultiline &&
            psiAtOffset.textOffset < caretOffset) {
            splitString(editor, psiAtOffset, caretOffset, true)
            return Result.Stop
        }
        return Result.Continue
    }
}